src/fd.cpp

```diff
@@ -0,0 +1,213 @@
+#include "fd.h"
+#include <Rcpp.h>
+#include <unistd.h>
+#include <cstdlib>
+#include <atomic>
+#include <memory>
+#include "tinycthread.h"
+#include "later.h"
+#include "callback_registry_table.h"
+
+extern CallbackRegistryTable callbackRegistryTable;
+
+class ThreadArgs {
+public:
+  ThreadArgs(
+    int num_fds,
+    struct pollfd *fds,
+    double timeout,
+    int loop
+  )
+    : timeout(createTimestamp(timeout)),
+      flag(std::make_shared<std::atomic<bool>>(false)),
+      fds(std::vector<struct pollfd>(fds, fds + num_fds)),
+      results(std::vector<int>(num_fds)),
+      num_fds(num_fds),
+      loop(loop) {}
+
+  ThreadArgs(
+    const Rcpp::Function& func,
+    int num_fds,
+    struct pollfd *fds,
+    double timeout,
+    int loop
+  ) : ThreadArgs(num_fds, fds, timeout, loop) {
+    callback = std::unique_ptr<Rcpp::Function>(new Rcpp::Function(func));
+  }
+
+  ThreadArgs(
+    void (*func)(int *, void *),
+    void *data,
+    int num_fds,
+    struct pollfd *fds,
+    double timeout,
+    int loop
+  ) : ThreadArgs(num_fds, fds, timeout, loop) {
+    callback_native = std::bind(func, std::placeholders::_1, data);
+  }
+
+  Timestamp timeout;
+  std::shared_ptr<std::atomic<bool>> flag;
+  std::unique_ptr<Rcpp::Function> callback = nullptr;
+  std::function<void (int *)> callback_native = nullptr;
+  std::vector<struct pollfd> fds;
+  std::vector<int> results;
+  const int num_fds;
```

Comment:
This seems redundant to `fds.size()` now

Reply:
Thanks. `num_fds` removed in [1e9107c](https://github.com/r-lib/later/pull/190/commits/1e9107c29060b3215814270dd3ff1666f4082c32).